bin/all_sky_search/pycbc_bin_trigger_rates_dq

```diff
@@ -103,108 +109,103 @@ logging.info("Applying %s > %.3f cut", args.sngl_ranking,
+                                       segment_name=flag_name,
+                                       ifo=ifo)
+
+# construct gate segments
+gating_segs = segmentlist([])
+if args.gating_windows:
+    gating_windows = args.gating_windows[ifo].split(',')
+    gate_before = float(gating_windows[0])
+    gate_after = float(gating_windows[1])
+    if gate_before > 0 or gate_after < 0:
+        raise ValueError("Gating window values must be negative "
+                         "before gates and positive after gates.")
+    if not (gate_before == 0 and gate_after == 0):
+        gating_segs = start_end_to_segments(
+                gate_times + gate_before,
+                gate_times + gate_after
+        ).coalesce()
+
+# make segments into mutually exclusive dq states
+gating_segs = gating_segs & analysis_segs
+flag_segs = flag_segs & analysis_segs
+
+dq_state_segs_dict = {}
+dq_state_segs_dict[2] = gating_segs
+dq_state_segs_dict[1] = flag_segs - gating_segs
+dq_state_segs_dict[0] = analysis_segs - flag_segs - gating_segs
+
+
+# utility function to get the dq state at a given time
+def dq_state_at_time(t):
+    for state, segs in dq_state_segs_dict.items():
+        if t in segs:
+            return state
+    return None
+
+
+# compute and save results
 with h5.File(args.output_file, 'w') as f:
-    for bin_name in locs_names:
-        bin_locs = locs_dict[bin_name]
-        inbin = np.isin(tmplt_ids, bin_locs)
-        trig_times_bin = trig_times_int[inbin]
-        trig_percentile = np.array([dq_percentiles_time[t]
-                                    for t in trig_times_bin])
-        logging.info('Processing %d triggers in bin %s',
-                     len(trig_percentile), bin_name)
-
-        (counts, bins) = np.histogram(trig_percentile, bins=(percentiles)/100)
-        counts = counts.astype(np.float64)
-        rates = np.zeros_like(bin_times)
-        rates[times_nz] = counts[times_nz]/bin_times[times_nz]
-        mean_rate = len(trig_percentile) / full_time
-        if mean_rate > 0.:
-            rates = rates / mean_rate
-
-        logging.info('Writing rates to output file %s', args.output_file)
-        grp = f.create_group(bin_name)
-        grp['rates'] = rates
-        grp['locs'] = locs_dict[bin_name]
-
-    f.attrs['names'] = locs_names
+    ifo_grp = f.create_group(ifo)
+    all_bin_grp = ifo_grp.create_group('bins')
+    all_dq_grp = ifo_grp.create_group('dq_segments')
+
+    # setup data for each template bin
+    for bin_name, bin_tids in bin_tids_dict.items():
+        bin_grp = all_bin_grp.create_group(bin_name)
+        bin_grp['tids'] = bin_tids
+
+        # get the dq states of the triggers in this bin
+        inbin = np.isin(tmplt_ids, bin_tids)
+        trig_times_bin = trig_times[inbin]
+        trig_states = np.array([dq_state_at_time(t) for t in trig_times_bin])
+
+        # calculate the dq rates in this bin
+        dq_rates = np.zeros(3, dtype=np.float64)
+        for state in [0, 1, 2]:
```

Comment:
I think it's nicer to do `for state, segs in dq_state_segs_dict.items()` and then `frac_dt = segs.abs() / livetime` as per my other comment